File: Algorithms/134. Gas Station/kotlin/solution.kt
class Solution {
    fun canCompleteCircuit(gas: IntArray, cost: IntArray): Int {
        var totalTank = 0
        var currTank = 0
        var startingStation = 0

        for (i in gas.indices) {
            totalTank += gas[i] - cost[i]
            currTank += gas[i] - cost[i]

            if (currTank < 0) {
                startingStation = i + 1
                currTank = 0
            }
        }

        if (totalTank >= 0) return startingStation
        return -1
    }
}

File: Algorithms/134. Gas Station/python/solution.py
class Solution:
    def canCompleteCircuit(self, gas: list[int], cost: list[int]) -> int:
        """
        :type gas: list[int]
        :type cost: list[int]
        :rtype: int
        """
        total_tank, curr_tank = 0, 0
        starting_station = 0
        for i in range(len(gas)):
            total_tank += gas[i] - cost[i]
            curr_tank += gas[i] - cost[i]
            # If one couldn't get here,
            if curr_tank < 0:
                # Pick up the next station as the starting one.
                starting_station = i + 1
                # Start with an empty tank.
                curr_tank = 0

        return starting_station if total_tank >= 0 else -1
